### DISCUSSION TITLE
Features for the Graph View

### DISCUSSION THREAD

**Discussion by nathanlesage** [Zettlr/Zettlr#3220] [Ideas]
[[EDITED]]

Today, Zettlr 2.2.2 shipped with the new graph view. As I made clear in the changelog, **the current implementation is still in its infancy**. That being said: In the future I plan on adding more features depending on what works, and what doesn't.

So one word of caution: **If you would like to have a feature for the graph view, please discuss it here and DO NOT OPEN AN ISSUE for this**. This is meant to prevent the issue tracker from flooding with numerous different feature requests. **Please utilize the tree-like structure of comments in this thread to try to keep it a little bit ordered :)**

## Already Planned

There are a few features I already have planned, please refrain from opening issues for these. I am planning on playing around with a lot of different ways of analyzing and visualizing networks since I have to do that for my PhD as well. I'll also consult faculty at my institute to make the graph view scientifically correct.

- [ ] Allow vertices to be scaled according to various node centrality measures (eigenvector centrality, betweenness centrality, degree centrality)
- [ ] Filter nodes by path/folder/title/tags (implemented with path, not yet tags)
- [x] Center the view on specific nodes (implemented)
- [x] Highlight specific nodes without completely hiding the others
- [ ] Add arrow heads to the links between nodes to visualize the directionality of the arcs
- [ ] Currently the community detection uses the weak criterion for clustering components; I plan to also add the strong community detection criterion (which will identify different components even if they are connected, if there is one link which is asymmetrical, and not mutual)
- [ ] Implement edge centrality
- [ ] Implement HITS (Hubs/Authority detection algorithm)
- [ ] Weighing of edges, for example utilizing edge betweenness centrality
- [ ] Different color shading of the links between nodes based on them being either asymmetrical or mutual (`MAN` criterion)
- [ ] Detection of roots and leafs
- [ ] Possibly integrate a better layouting mechanism such as Fruchterman-Reingold or MDS-based layouts (e.g., Kamada/Kawai)
- [ ] Clicking on a vertex should highlight the path from its parent to all leaf nodes

## Known Issues

Also, there are a few known issues that I will be gradually improving, so again: Please do not open an issue for these as well!

* Dragging the network graph does not yet work
- [x] Currently the graph will rebuild itself even if nothing in the link structure has changed (i.e. by opening or saving a file) (fixed)
- [ ] Links overlap each other; and in general the simulation is somewhat awkward (I couldn't find any solutions for this yet)
- [ ] The graph building process still takes a lot of time, depending on how many links you have, there is lots of room for optimization.

**Comment by xauximan** [comment]
At first I would like to thank you for your work. I really like to work with Zettlr for more than one and a half years and I also used it for my first state examination for teaching geography and social studies. Works even great under immense time constrains!

So here are some of my ideas for possible features of the graph view: 
**1. Switch for displayed filename**
I think it would be really helpful to have a switch for the name which is displayed in the graph view, so you are able to switch from file name to first header and vice versa. Maybe it's also possible to get filename from YAML-header. 

**2. Using scroll wheel for zooming in and out**
I think this depends on your point of view. I think zooming in and out with scroll wheel feels more natural. On top of that I see no need for scrolling upwarts and downwarts, if you are able to simply zoom in and out the parts of the graph you are interested in.

**3. Possibility to visualise Folgezettel**
That could be helpful for people working like Niklas Luhmann. Personally I dropped that technique and I am using Structure Zettels because it's easier to store Zettels in various lines of reasoning inside structure zettels. But it might be possible that some people out there using it. Niklas Luhmann Archive shows that visualization of Folgezettel is basicly possible. 

**4. Possibility to choose folders and adjust forces**
Limiting the graph to one or more folders might be useful for those who are working with a bunch of folders to avoid a mess and keeping Zettelkasten accessible. Also it can help to reduce graph view and to get a better overwiew about a specific topic or problem. Adjustible forces like in Obsidian might help to handle the chaotic nature of Luhmann-like Zettelkasten likewise. Furthermore it helps working with huge amounts of files if you are working with the principle of atomicy. 

**5. Show non existing files and tags**
This can also create new insights and it would be nice if there are both options. Some people working with tags like [[tag]]. Further showing non existing files can point you on files, where you did a mistake.

**Reply by kyaso** [reply]
> 5. Show non existing files and tags
> This can also create new insights and it would be nice if there are both options. Some people working with tags like [[tag]]. Further showing non existing files can point you on files, where you did a mistake.

I agree with this one. I also use these `[[tag]]` tags. In my workflow, these are different from regular `#tag` tags, in the sense that the latter are attached to the _entire_ note (and, ideally, only appear _once_ in the file), while the former I can use _multiple_ times in the same file and also in other files. So, when two files share the same `[[tag]]`, they are linked. I could imagine visualizing these kinds of links as creating a "virtual" node for the `[[tag]]`, and then connecting each note that contains that tag to that virtual node (so it would be a star topology).

**Reply by davidoskky** [reply]
> 4. Possibility to choose folders and adjust forces

It would also be great to have an option to move a cluster to a specific directory.
I, for example, take notes regarding new subjects in one directory and if I start writing a lot of notes about it I'll move all of them to a dedicated directory.
Selecting to move a cluster could ease up this process, as long as all the notes belonging to the same subject are connected among them and isolated from other stuff.

**Reply by nathanlesage** [reply]
With regard to non existing files: This is already happening, the graph also shows non-existing files.

**Comment by davidoskky** [comment]
Allow custom coloring of the notes.
If I have a large cluster of notes it would be nice to quickly identify the ones which belong to a particular tag, or to a search term by coloring.
This would make it easier to understand the graph and maybe give insights on connections that should be removed or added.

**Comment by mousetrappling** [comment]
I'm not sure how big a file set you've tested it on, so maybe some statistics are helpful for figuring out how to optimise this.

I have 2374 files, not sure how many links per file but probably 5 to 10 on most with some with in excess of 100 (index or outline files).   I gave up waiting after 17 minutes, when it was saying it had processed ~1100 files.

The CPU in my computer is: Intel Quad-core i7-1165G7 (2.8GHz, 4.7GHz Turbo), and the machine has 32Gb of ram.  My OS is Ubuntu 20.04.  I looked at "top" while Zettlr was working on the graph, and the process was using 100% CPU (which I think means it had maxed out one core not that it had maxed out the whole CPU), it was barely using any memory.

**Reply by nathanlesage** [reply]
> it was barely using any memory.

Well, at least *one* optimization problem I solved. But I agree, building the graph is currently utterly un-optimized, I have to improve that definitely.

**Reply by nathanlesage** [reply]
Just found an easy way to massively improve graph build times; I'm looking forward to hearing your verdict on the next version once it's out!

**Reply by mousetrappling** [reply]
That certainly worked ... with 2.2.3 it took 3.5s to build (according to the logs)!

**Comment by halloichbingunnar** [comment]
Sorry for opening up an issue instead of posting here – but anyways, the new Graph View is fantastic! 

I just wondered if it would be possible to somehow include tag similarity in the graph view, as it is already included in the sidebar. I think this could be a bit challenging to implement, but nevertheless very insightful when drawing new connections between zettels and notes. Maybe here someone with more experience in graphical visualisation may jump in, but this is how it popped up in my mind:

There could be a toggle to use links or tags in the graph (I think we can't have both at the same time?). 

In the tag mode, it could look like this:

* Tags appear as simple dots, maybe differently shaped
* The graph displays the connections of file-dots to the tag-dots. just like in the link-graph

**Reply by xauximan** [reply]
In obsidian tags are graphically similar to files. Each tag is a new dot with lines to the files where the tag is used. And this is the best way I can imagine. My advice (after one semester of cartography) is to use different colours for tags instead of different shape, because colours are more selective than shapes in human perception. So in a nutshell: you see the difference better
I think your idea to switch between links and tags in the graph is quiet interesting ad worth a try. Never had seen something like that before. Both at once seems to be no problem, because I have seen this often.

**Reply by nathanlesage** [reply]
> possible to somehow include tag similarity in the graph view, as it is already included in the sidebar.

Yes, and this is one of the few arc-attributes I plan to implement. But I have to figure out ways of doing so good, so give it some time!

**Comment by s25roy** [comment]
I updated Zettlr to 2.2.2 to check graph view out immediately hunting for a button or option somewhere to enable it but couldn't find any such. Should be noted that I only finished the basic tutorial so there are very few links. Checked my preferences, read the docs on [graph view](https://docs.zettlr.com/en/academic/graph/) but just totally beyond me where the option is in the UI, this could be a really dumb question am I missing something obvious? 

P.S. Just started using Zettlr and probably first post on Github so forgive if this isn't exactly the right place/way to post this.

**Reply by wheelercj** [reply]
Hi! I put some photos with red circles showing what to click on to bring up the graph view [here](https://wheelercj.github.io/notes/pages/20220217110602.html). Since the feature is so new, I wouldn't be surprised if this becomes easier in the future.

**Reply by s25roy** [reply]
@wheelercj Thanks for the illustration, I didn't intuitively figure it would be under stats>more stats, might have eventually but this saved me a lot of time. @nathanlesage wonder if it would be nice in terms of UI to have a dedicated icon/button next to tags or something for graph view? 

I've been trying to follow some of the debate of the usefulness of graphs/networks in the first place in your blog posts and I understand the concerns with how network view representations can be misleading. Perhaps it is intentional to not bring extra attention with a dedicated button also keeping the UI minimalist in its philosophy but for a new user at least, it is still difficult to find.

**Reply by wheelercj** [reply]
I agree and it looks like the plan is to keep the graph feature slightly hidden while it's brand new since there could be a flood of duplicate feature requests and bug reports. Then when the feature has matured a bit, the button will be moved. I'm kind of guessing about this though.

**Comment by StoltHD** [comment]
A few "wishes"

1. Colorize Nodes based on "type" or other Yaml keypair

2. Possible to "cluster" or group nodes based on keywords or tags (differ from the normal "nearest neighbors linked nodes)

3. most people will do their graph configuration ones and maybe just add to it or do small tweaks, so please give us the opportunity to configure the graph settings in a config file (like json or something like that), not only as a bloated panel that can be a mile long if you have a lot of configurations.

4. Possible to save the graph layout to a network graph file format like gefx, graphml, cytoscape json, etc. that can be open in any network graph software for extended analyzes and include any yaml keys as attributes/properties and the body of the note as a note... e.g. interchangeability to network graph software both desktop and web...
(It would be easy to make a library of your complete note library to use in a software like e.g. openmappr etc.)

5. It would be great if it would be possible to set labels on edges, it could be done e.g. with a yaml keypair in the target note, e.g. "type: friend", the label will be written in the configuration file something like "if type=friend, label=friend.

6. Node names based on yaml keypair e.g. "alias", "name" etc. in addition to filename and "first header"

I really like that you have the graph in a separate window.
Maybe it will be possible to use multi-thread and Opengl for faster calculation of the graph?

**Reply by nathanlesage** [reply]
> Maybe it will be possible to use multi-thread and Opengl for faster calculation of the graph?

OpenGL is already being used (by Chromium in the background). The slow graph building was just because I ran the community detection algorithm every time a new tie was created, which was not optimal. Fixed that already, and coming with the next update

**Comment by mousetrappling** [comment]
I've gone back and forth over the last couple of years on whether a graph view would be useful (as opposed to pretty), so I'm not the best person to give feedback/suggest features as I was coming down on the side of "maybe pretty but that's all".

However, even though I was thinking it probably wouldn't be useful I guess I expected something a little more informative than I am getting, see [screenshot](http://ninecats.org/margaret/temp/Screenshot%20from%202022-02-19%2015-52-09.png).  It does tell me that I've only really got one big cluster of most of my ~2.5k files, plus a small handful of isolates and isolated mini-clusters, so maybe that's why it feels less informative.

Maybe it would be interesting if I could highlight a node and have all the ones that it links to light up, optionally two or three levels deep?

Also I'm not sure what the purpose of zooming in is?  I had assumed the spaces between nodes would get bigger so I could see the connections, or something like that, but when I zoom in with the mousewheel I get a screenful of densely packed big dots! [Screenshot](http://ninecats.org/margaret/temp/Screenshot%20from%202022-02-19%2015-52-30.png) Have I missed a setting or way to use it?

One thing that would definitely be nice would be to be able to see the yaml title or H1 rather than the filename for a node, my filenames are ids which is uninformative without opening the file.

**Reply by nathanlesage** [reply]
> Maybe it would be interesting if I could highlight a node and have all the ones that it links to light up, optionally two or three levels deep?

Sounds like a reasonable request.

> One thing that would definitely be nice would be to be able to see the yaml title or H1 rather than the filename for a node, my filenames are ids which is uninformative without opening the file.

Absolutely, and this is already on the roadmap!

**Comment by jcaonr** [comment]
The graph view is a great addition to zettlr. Thx for developing such a great tool.

When I'm working in Zettlr I'm using the preferences option 'Display files using - First heading level 1 if applicable' and my filenames on harddrive are rather cryptic. I would suggest an option to show the labels in graph view corresponding to the set 'Display files using -  ...' option. So that in my case for example the first heading is shown as label in graph view.

**Comment by trueschool** [comment]
Thanks for paying attention to this important feature of Zettlr! I'm currently using Obsidian on a work project mostly for its graph feature's robustness, so I'd love to see Zettlr catch up in this respect. 

As for features I'd like to see: 
- I'd like it if the "components" drop down were a bit more descriptive, taking names from specific folders in the workspace. For example, "Component 6" in the current example isn't descriptive of the content to the user. 

![image](https://user-images.githubusercontent.com/183095/158217795-9ad82c6b-c561-44ce-a759-917a1fde8d16.png)

- As stated above, being able to see tags on the nodes would be huge. 

Thanks again for considering!

**Reply by datavectors** [reply]
Is it feasible for an external automation script to access the "GraphView internals" in say json or other format? Where is GraphView saved in filesystem?
The reason is that I am experimenting with Zettlr in a toolchain.
Typically I might try displaying the nodes network using D3.js.

[Background] I am migrating from Atom + Markdown Preview Enhanced to try Zettlr following GitHub's decision to "sunset" Atom with all its "hackability" features.  My current intention is to use Zettlr in a toolchain which passes data to other tools. One such "hack" might be to expand GraphView into a D3.js rendering in browser. I have read some replies to feature requests and I conclude it will be easier to write separate helper scripts myself.

**Reply by nathanlesage** [reply]
Well the graph view is actually displayed using d3, but it is always computed and never stored to disk...

**Reply by datavectors** [reply]
Further digging revealed ~/.config/Zettlr/config.json
I am now using Python to scrape this.   openPaths and openFiles.
I can offer a simple Python script for this.
Then I will try building something like this:

https://philogb.github.io/jit/static/v20/Jit/Examples/ForceDirected/example1.html

I plan to analyse the documents tree (vault?) to then apply concordance analysis of all documents (the corpus of documents) using companion linguistics tools such as AntConc.
https://www.laurenceanthony.net/software/antconc/

This one case study illustrates the advantages of using Zettlr in a toolchain, cooperating with other applications and frameworks. This is why I advocate automation scripts.

**Comment by aljohnston112** [comment]
Would it be possible to get a graph view that uses tags as links?

**Reply by nathanlesage** [reply]
Possible, yes, I don't know how difficult that may be, though.
